File: smolyak.py
"""
This file contains a class that builds a Smolyak Grid.  The hope is that
it will eventually contain the interpolation routines necessary so that
the given some data, this class can build a grid and use the Chebyshev
polynomials to interpolate and approximate the data.

Method based on Malin, Krueger, Kubler 2010 and
                Judd, Maliar, Maliar, Valero 2013 (W.P)

Author: Chase Coleman
"""

import numpy as np
import numpy.linalg as la
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
from itertools import product
import time as time

class Smolyak(object):
    """
    This class currently takes a dimension and a degree of polynomial
    and builds the Smolyak Sparse grid.

    Attributes:
    ------------------
    d : scalar : integer
        This is the dimension of grid that you are building

    mu : scalar : integer
        mu is a parameter that defines the fineness of grid that we
        want to build


    Methods:
    --------
    grid_build : This method builds the sparse grid

    """
    def __init__(self, d, mu):
        """
        Parameters
        ----------
        d : scalar : integer
            This is the dimension of grid that you are building

        mu : scalar : integer
            mu is a parameter that defines the fineness of grid that we
            want to build
        """
        self.d = d
        self.mu = mu

        if mu<1:
            raise ValueError('The parameter mu needs to be > 1.')

        if d<=1:
            raise ValueError('You are trying to build a one dimensional\
                             grid.')

    def _m_seq(self, vals):
        """
        Pass in a value and it will calculate the ith element of the
        sequence of m's as described in paper.
        """

        # Create vec of ones in same shape as vals.
        # Everywhere vals=1 we want to put a 1, so saves a step.
        ret_vals = np.ones(vals.shape)

        # Everywhere else it should be 2**(i-1) + 1
        ret_vals[np.where(vals > 1)] = 2**(vals[np.where(vals > 1)] - 1) + 1

        return ret_vals

    def _g_func(self, n):
        """
        This function returns the set G.  It corresponds to the
        Chebyshev extrema.
        """

        if abs(n - 1) < 1e-10:
            g_set = np.array([0])
        else:
            g_set = np.cos(np.pi*(np.arange(n))/(n-1.))
            # make sure they are 0 where I want them 0
            g_set[np.where(abs(g_set) < 1e-14)] = 0.

        return g_set

    def find_points(self):
        """
        For integers between 1 and d find the sets of d integers that
        add to mu + d.

        i.e. mu=1, d=3
        {1, 1, 2}, {1, 2, 1}, {2, 1, 1}
        """

        d = self.d
        mu = self.mu

        # We want all possible values up to dimension + 1
        poss_vals = np.arange(mu + 1) + 1

        # Create empty list to fill with the points
        points = []

        # Use this for loop to create the list of points
        for element in product(poss_vals, repeat=d):
            # We only want the ones that sum to d + mu
            if sum(element) == d + mu:

                # Transform the element i to m(i)
                elements = self._m_seq(np.array(element))

                # Now take these points and make the set g
                temp_points = map(self._g_func, elements)

                # Put those points into the list of points
                points.append(list(product(*temp_points)))

        return points


    def build_grid(self):
        """
        This method builds a d dimensional grid to be used for
        interpolation.
        """
        d = self.d
        mu = self.mu

        # Call the find_points function to find the elements that we
        # need to build the grid
        points = self.find_points()
        point_set = set()

        # Put each of the points into a set so that we don't have
        # duplicate points
        for elem in points:
            for temp in elem:
                point_set.add(temp)

        # Now put them in an array.  Each row is a d-dimensional vector
        # that represents the point.
        point_array = np.array(list(point_set))

        self.grid = point_array

#----------------------------------------------------------------------#
# Example
#----------------------------------------------------------------------#

def check_points(d, mu):
    if abs(mu - 1) < 1e-14:
        return 2*d + 1

    if abs(mu - 2) < 1e-14:
        return 1 + 4*d + 4*d*(d-1)/2.

    if abs(mu - 3) < 1e-14:
        return 1 + 8*d + 12*d*(d-1)/2. + 8*d*(d-1)*(d-2)/6.


d = 3
mu = 10
# start = time.time()
smolyak_grid = Smolyak(d, mu)
smolyak_grid.build_grid()
# elapsed = start - time.time()
# point_array = smolyak_grid.grid

# print(smolyak_grid.grid.shape)
# print(check_points(d, mu), "Took %.5f" %elapsed)


# Plot to show the where points go for 3-D case

# fig = plt.figure(1)
# fig.suptitle("Smolyak Grid on 3 Dimensions with $\mu =$ %d"  %smolyak_grid.mu)
# ax1 = fig.add_subplot(221)
# ax1.set_title("X-Y Grid")
# ax1.plot(point_array[:, 0], point_array[:, 1], 'ko')
# ax1.set_ylim((-1.5, 1.5))
# ax1.set_xlim((-1.5, 1.5))
# ax2 = fig.add_subplot(222)
# ax2.set_title("Y-Z Grid")
# ax2.plot(point_array[:, 1], point_array[:, 2], 'ko')
# ax2.set_ylim((-1.5, 1.5))
# ax2.set_xlim((-1.5, 1.5))
# ax3 = fig.add_subplot(223)
# ax3.set_title("X-Z Grid")
# ax3.plot(point_array[:, 0], point_array[:, 2], 'ko')
# ax3.set_ylim((-1.5, 1.5))
# ax3.set_xlim((-1.5, 1.5))
# ax4 = fig.add_subplot(224, projection='3d')
# ax4.set_title("3-D Projection")
# surf = ax4.scatter(point_array[:, 0], point_array[:, 1], point_array[:, 2])
# ax4.set_ylim((-1.5, 1.5))
# ax4.set_xlim((-1.5, 1.5))
# ax4.set_zlim((-1.5, 1.5))
# plt.show()

File: smolyak.jl
using Iterators
using Cartesian
import PyPlot

choose(n, k) = factorial(n) / (factorial(k) * factorial(n - k))

count_coef(d, mu, i) = (-1) ^ (d + mu - i) * choose(d - 1, d + mu - i)

# takes the ith slice of A in the dth dimension and squeezes along d
slice_sqz(A, d, i) = squeeze(slicedim(A, d, i), d)


function m_i(i::Int)
    if i < 1
        error("DomainError: i must be positive")

    elseif i == 1
        return 1
    else
        return 2^(i - 1) + 1
    end
end


function cheby2n(x, n::Int; kind=1)
    # Evaluates the first n+1 Chebyshev polynomials of the 'kind' kind at x
    # NOTE: This will only work when kind = 1 or kind = 2
    # NOTE: We evaluate the first n+1, with the first dimension being all 1's
    dim = size(x)
    results = zeros(n + 1, dim...)
    results[1, :] = 1.
    results[2, :] = kind * x
    for i=3:n+1
        results[i, :] = (2 * x .* slice_sqz(results, 1, i-1) -
                         slice_sqz(results, 1, i-2))
    end
    return results
end


function S_n(n::Int)
    # Compute the set S_n. All points are extrema of Chebyshev polynomials

    if n < 1
        error("DomainError: n must be positive")
    end

    if n == 1
        return [0.]
    end

    m = m_i(n)
    j = [1:m]
    pts = cos(pi * (j - 1.) / (m - 1))::Array{Float64, 1}
    for i = 1:size(pts, 1)
        pts[i] = abs(pts[i]) < 1e-12 ? 0.0 : pts[i]
    end
    return pts
end


function A_n(n::Int)
    if n < 1
        error("DomainError: n must be positive")
    end

    if n == 1
        return [0.]
    elseif n == 2
        return [-1., 1.]
    end

    sn = S_n(n)
    return sn[[2:2:size(sn, 1)]]
end


function A_chain(n::Int)
    # Return Dict of all A_n from n to 1. Keys are n, values are arrays

    sn = S_n(n)

    a = Dict{Int, Array{Float64, 1}}()
    sizehint(a, n)

    # These are constant and don't follow the pattern.
    a[1] = [0.]
    a[2] = [-1., 1.]

    for i=n:-1:3
        # push!(a, sn[[2:2:size(sn, 1)]])
        a[i] = sn[[2:2:size(sn, 1)]]
        sn = sn[[1:2:size(sn, 1)]]
    end

    return a
end


function dense_grid(d::Int, mu::Int)
    # Use nested Smolyak sets to construct Smolyak grid

    p_vals = [1:mu+1]

    # TODO: This can probably be optimized to not be of type Any
    points = Any[]

    # Smolyak.find_points(self)
    for el in product([p_vals for i in [1:d]]...)
        if sum(el) == d + mu
            temp_points = map(S_n, el)
            append!(points, [collect(product(temp_points...))...])
        end
    end

    # Build Smolyak grid
    p_set = Set()

    for elem in points
        push!(p_set, [elem...])
    end

    unique_pts = collect(p_set)
    n_pts = size(unique_pts, 1)
    grid = Array(Float64, n_pts, d)
    for i=1:n_pts
        grid[i, :] = unique_pts[i]
    end

    return grid
end


function num_grid_pts(d::Int, mu::Int)

    if mu == 1
        return int(2 * s - 1)
    elseif mu == 2
        return int(1 + 4*d + 4*d*(d-1)/2)
    else
        return int(1 + 8*d + 12*d*(d-1)/2 + 8*d*(d-1)*(d-2)/6)
    end

end

function sparse_grid(d::Int, mu::Int)
    # Use disjoint Smolyak sets to construct Smolyak grid

    p_vals = ones(Int64, d)*(mu+1)
    An = A_chain(mu + d)

    # TODO: This can probably be optimized to not be of type Any
    points = Any[]

    n_pts = num_grid_pts(d, mu)::Int
    grid = Array(Float64, n_pts, d)
    n = 1

    @forcartesian el p_vals begin
        if d <= sum(el) <= d + mu
            temp_points = [An[i] for i in el]
            for pt in product(temp_points...)
                grid[n, :] = [pt...]
                n += 1
            end
            # append!(points, [collect(product(temp_points...))...])
        end
    end

    # n_pts = size(points, 1)::Int
    # grid = Array(Float64, n_pts, d)
    # for i=1:n_pts
    #     grid[i, :] = [points[i]...]
    # end

    return grid
end


function plot_grid(g)
    if size(g, 2) == 2
        PyPlot.plot(g[:, 1], g[:, 2], "b.")
        PyPlot.xlim((-1.5, 1.5))
        PyPlot.ylim((-1.5, 1.5))
    elseif size(g, 2) == 3
        PyPlot.scatter3D(g[:, 1], g[:, 2], g[:, 3], "b.")
    else
        throw("ERROR: can only plot 2d or 3d grids")
    end
end
